man/man1/pmquery.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.TH PMQUERY 1 "" "Performance Co-Pilot"
.SH NAME
\f3pmconfirm\f1,
\f3pmmessage\f1,
\f3pmquery\f1 \- general purpose dialog box
.SH SYNOPSIS
\f3pmconfirm\f1
[\f3\-c\f1]
[\f3\-b\f1 \f2button-name\f1]
[\f3\-B\f1 \f2default-button-name\f1]
[\f3\-t\f1 \f2string\f1]
[\f3\-file\f1 \f2filename\f1]
[\f3\-icon\f1 \f2icontype\f1]
[\f3\-font\f1 \f2font\f1]
[\f3\-header\f1 \f2titlebar-string\f1]
[\f3\-useslider\f1]
[\f3\-noslider\f1]
[\f3\-noframe\f1]
[\f3\-exclusive\f1]
.br
.PP
\f3pmmessage\f1
[\f3\-buttons\f1 \f2label1\f1[:\f2value1\f1][,\f2label2\f1[:\f2value2\f1][,...]]]
[\f3\-center\f1]
[\f3\-nearmouse\f1]
[\f3\-default\f1 \f2button\f1]
[\f3\-file\f1 \f2filename\f1]
[\f3\-print\f1]
[\f3\-timeout\f1 \f2sec\f1]
.I message...
.br
.PP
\f3pmquery\f1
[\f3\-input\f1]
[\f3all above options...\f1]
[\f2message...\f1]
.SH DESCRIPTION
.B pmquery
provides a command-line-option compatible implementation of the
.B xconfirm
and
.B xmessage
tools, using a look-and-feel that is consistent with
.BR pmchart .
Several extensions to the functionality of the original tools have been made,
in order to improve their specific utility for
.BR pmchart ,
but wherever possible the original semantics remain.
.PP
.B pmconfirm
displays a line of text for each
.B \-t
option specified (or a file when the
.B \-file
option is used),
and a button for each
.B \-b
option specified.
When one of the buttons is pressed, the label of that button is written to
.B pmquery's
standard output.
This provides a means of communication/feedback from within shell
scripts and a means to display useful information to a user from
an application.
.PP
.B pmmessage
displays a window containing a message from the command line, a file,
or standard input.
It additionally allows buttons to be associated with an exit status,
and only optionally will write the label of the button to standard output.
.PP
.B pmquery
extends the above tools to additionally support limited user input,
as free form text.
In this
.B \-input
mode, any text entered will be output when the default button is pressed.
A default text can be entered using the same mechanisms as the other tools.
.PP
Command line options are available to specify font style, frame style,
modality and one of several different icons to be presented for tailored
visual feedback to the user.
.SH OPTIONS
The available command line options are:
.TP 5
.B \-c\f1, \f3\-center\f1
Center the window on the display.
.TP
.B \-nearmouse
Pop up the window near the mouse cursor.
.TP
.B \-b \f2button-name\f1
Displays a button with the label
.IR button-name .
If
.I button-name
is the empty string, the button in that position is not displayed.
If no
.B \-b
options are present, the default is a button with the label Continue.
The exit status associated with
.I button-name
is zero.
.TP
.B \-B \f2button-name\f1
Displays a button with the label
.I button-name
and specifies it as the button to be activated when enter is pressed.
The exit status associated with
.I button-name
is zero.
.TP
.B \-buttons \f2button,button,.\|.\|.\f1
This option will create one button for each comma-separated \f2button\f1
option.
Each \f2button\f1 consists of a label optionally followed by a colon
and an exit value.
The exit value will be returned if that button is selected.
The default exit value is 100 plus the button number.
Buttons are numbered from the left starting with one.
.TP
.B \-default \fIlabel\fP
Defines the button with a matching \fIlabel\fP to be the default.
If not specified there is no default.
The corresponding resource is \fBdefaultButton\fP.
Pressing Return anywhere in the \fIxmessage\fP window will activate
the default button.
The default button has a wider border than the others.
.TP
.B \-t \f2message\f1
Displays message.
Any number of strings can be listed on the command line
(each must be preceded with the
.B \-t
option).
.TP
.B \-file \f2filename\f1
Displays the file
.I filename.
All
.B \-t
options will be ignored.
A \f2filename\f1 of `\f2\-\f1' reads from standard input.
.TP
.B \-icon \f2icontype\f1
Displays the icon
.I icontype
where icontype is one of:
.IR info ,
.IR error ,
.IR question ,
.IR warning ,
.IR critical .
.I action
is also accepted as a synonym for
.I error
for backward compatibility.
.BR pmquery
introduces the additional
.I archive
and
.I host
icon types as well as the original
.BR xconfirm
types listed earlier.
.TP
.B \-font \f2fontname\f1
Use fontname as the font.
This option is only available when using the X Window System.
.TP
.B \-header \f2string\f1
Use string as the window title.
.TP
.B \-print
This causes the program to write the label of the button pressed to
standard output.
It is the default behaviour for
.B pmconfirm
and
.BR pmquery .
.TP
.B \-noprint
This causes the program to not write the label of the button pressed to
standard output.
It is the default behaviour for
.BR pmmessage .
.TP
.B \-geometry \f2geometry-string\f1
This provides the tool with an X-compatible geometry string specification.
This option is only available when using the X Window System.
.TP
.B \-useslider
When displaying a file, always use a slider instead of determining
automatically whether a slider is necessary.
.TP
.B \-noslider
Do not create a slider, and clip text to the window size, instead of
determining automatically whether a slider is necessary..
.TP
.B \-noframe
Do not display a frame around the contents.
.TP
.B \-exclusive
Grab the keyboard/pointer and do not allow further
input until a button is pressed.
.TP
.B \-timeout \f2secs\f1
Exit with status 0 after \fIsecs\fP seconds if the user has not
clicked on a button yet.
The corresponding resource is \fBtimeout\fP.
.TP
.B \-?\f1, \f3\-h\f1, \f3\-help\f1
Display usage message and exit.
.SH EXAMPLES
The following shell script will display a window with an information icon,
asking the user a yes or no question with "Yes" as the default.
.PP
.nf
 #! /bin/sh
 case `pmquery \-t "Really power down?" \-b No \-B Yes \-icon question
 in
   Yes) shutdown;;
   No) ;;
 esac
.fi
.PP
A second example, which prompts for a hostname then starts a
terminal with an ssh session connected to the requested host.
.PP
.nf
 #! /bin/sh
 host=`pmquery \-input \-icon host \-b Cancel \-B OK \\
               \-header "Remote Terminal \- Secure Shell"
 [ "$host" = "Cancel" \-o \-z "$host" ] && exit
 gnome-terminal \-e "ssh $host"
.fi
.SH ENVIRONMENT
.B pmquery
is an excellent choice of utility for the
.B PCP_XCONFIRM_PROG
Performance Co-Pilot configuration parameter (refer to
.BR pcp.conf (5)
for details).
.PP
Note that
.B PCP_XCONFIRM_PROG
will be automatically set to
.B pmquery
inside tools like
.BR pmchart ,
unless
.B PCP_XCONFIRM_PROG
is already set in the environment.
.SH EXIT STATUS
If it detects an error,
.B pmquery
always returns 1, so this value should not be associated with a button.
Unless \f2\-button\f1 option has not been used, the return code will be
zero on success.
.SH SEE ALSO
.BR pmchart (1)
and
.BR pcp.conf (5).

.\" control lines for scripts/man-spell
.\" +ok+ defaultButton esac fontname noprint pmquery's titlebar xmessage


CHANGELOG
pcp-6.3.6 (17 March 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.4 (13 March 2025)
    - libpcp: fix endian issue in big-endian __pmTimestamp writing
    - pcp-xsos: fine-tune error handling on bad command line options
    - pcp-xsos: single-line pmPrintValue/pminfo values, escape shell chars
    - pcp-atop: correct BPF per-process network metric names
    - pmcd, pmdaproc: allow non-local username connection attributes
    - pmlogger, libpcp: fix handling of dynamic instance domain labels
    - pmrep: use BPF instead of BCC for proc networking metrics
    - pmlogconf: use libbpf over BCC for networking metrics
    - pmlogextract: fix logic around instances labels, allowing set updates
    - pmdalinux: mem.numa.hugepages metrics (per node, per hugepage size)
    - pmdabpftrace: fix for some python3 versions async IO create_task
    - pmdanvidia: if present use dso from the libnvidia-ml runtime package
    - pmdadm, pmieconf: default to enabling the dmthin pmie rules
    - configure.ac: fix autoreconf, add AC_DEFINE_UNQUOTED descriptions
    - rpm builds: enable bpf, bpftrace, bcc on riscv64 builds
    - rpm builds: drop user creation scriptlets for rpm builds
    - deb build: force debian/rules to be recreated each time
    - containers: updates to the archive-analysis container
    - selinux: several policy updates for new AVC denials
    - docs: improve SSL/TLS diagnostic in pcp-verify script

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out

src/pmquery/pmquery.cpp
/*
 * Copyright (c) 2014,2023 Red Hat.
 * Copyright (c) 2007, Aconex.  All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 */
#include "pmquery.h"

#define max(a,b) ((a)>(b)?(a):(b))
#define DEFAULT_EDIT_WIDTH	640	/* in units of pixels */
#define DEFAULT_EDIT_HEIGHT	320	/* only for usesliderflag */

enum icontypes {
    INFO_ICON,
    ERROR_ICON,
    QUESTION_ICON,
    WARNING_ICON,
    ARCHIVE_ICON,
    HOST_ICON,
} iconic;

static const char *title = "Query";
static int timeout;
static int *statusi;
static const char **buttons;
static int buttoncount;
static const char *defaultbutton;
static char **messages;
static int messagecount;

static void nomem()
{
    fputs("Insufficient memory\n", stderr);
    exit(1);
}

int PmQuery::setTimeout(char *string)
{
    char *endnum;
    timeout = (int)strtol(string, &endnum, 10);
    if (*endnum != '\0' || timeout <= 0)
	return -1;
    return 0;
}

void PmQuery::setTitle(char *heading)
{
    title = heading;
}

int PmQuery::messageCount()
{
    return messagecount;
}

int PmQuery::buttonCount()
{
    return buttoncount;
}

int PmQuery::setIcontype(char *string)
{
    if (strcmp(string, "info") == 0)
	iconic = INFO_ICON;
    else if (strcmp(string, "error") == 0 ||
	strcmp(string, "action") == 0 ||
	strcmp(string, "critical") == 0)
	iconic = ERROR_ICON;
    else if (strcmp(string, "question") == 0)
	iconic = QUESTION_ICON;
    else if (strcmp(string, "warning") == 0)
	iconic = WARNING_ICON;
    else if (strcmp(string, "archive") == 0)
	iconic = ARCHIVE_ICON;
    else if (strcmp(string, "host") == 0)
	iconic = HOST_ICON;
    else
	return -1;
    return 0;
}

void PmQuery::addMessage(char *string)
{
    messages = (char **)realloc(messages, (messagecount+1) * sizeof(char *));
    if (!messages)
	nomem();
    messages[messagecount++] = string;
}

void PmQuery::addButton(const char *string, bool iamdefault, int status)
{
    buttons = (const char **)realloc(buttons, (buttoncount+1) * sizeof(char *));
    statusi = (int *)realloc(statusi, (buttoncount+1) * sizeof(int));
    if (!buttons)
	nomem();
    if (iamdefault)
	defaultbutton = string;
    statusi[buttoncount] = status;
    buttons[buttoncount++] = string;
}

void PmQuery::addButtons(char *string) // comma-separated label:exitcode string
{
    char *n;
    QString pairs(string);
    static int next = 100;

    QStringList list = pairs.split(",");
    for (QStringList::Iterator it = list.begin(); it != list.end(); ++it) {
	QString name = (*it).section(":", 0, 0);
	QString code = (*it).section(":", 1, 1);
	if (!name.isEmpty()) {
	    int sts = code.isEmpty() ? ++next : code.toInt();
	    if ((n = strdup(name.toLatin1().data())) == NULL)
		nomem();
	    addButton(n, false, sts);
	}
    }
}

void PmQuery::setDefaultButton(char *string)
{
    for (int i = 0; i < buttoncount; i++)
	if (strcmp(buttons[i], string) == 0)
	    defaultbutton = buttons[i];
}

void PmQuery::buttonClicked()
{
    done(my.status);
}

void PmQuery::timerEvent(QTimerEvent *)
{
    done(1);
}

// Currently we set default edit size to hardcoded values, until
// better ways are found to interact with any geometry requests.
// Note: the +4 pixels for height ensure the auto-scroll does not
// kick in, seems to be required.

PmQuery::PmQuery(QApplication &application,
		bool inputflag, bool printflag, bool noframeflag,
		 bool nosliderflag, bool usesliderflag, bool exclusiveflag)
    : QDialog(application.activeWindow())
{
    QHBoxLayout *hboxLayout;
    QVBoxLayout *vboxLayout;
    QSpacerItem *spacerItem;
    QSpacerItem *spacerItem1;
    QVBoxLayout *vboxLayout1;
    QHBoxLayout *hboxLayout1;
    QSpacerItem *spacerItem2;

    my.status = 0;

    QString filename;
    if (iconic == HOST_ICON)
	filename = tr(":images/dialog-host.png");
    else if (iconic == ERROR_ICON)
	filename = tr(":images/dialog-error.png");
    else if (iconic == WARNING_ICON)
	filename = tr(":images/dialog-warning.png");
    else if (iconic == ARCHIVE_ICON)
	filename = tr(":images/dialog-archive.png");
    else if (iconic == QUESTION_ICON)
	filename = tr(":images/dialog-question.png");
    else // (iconic == INFO_ICON)
	filename = tr(":images/dialog-information.png");

    QIcon	icon(filename);
    QPixmap	pixmap(filename);
    setWindowIcon(icon);
    setWindowTitle(tr(title));

    QGridLayout *gridLayout = new QGridLayout(this);
    gridLayout->setSpacing(6);
    gridLayout->setContentsMargins(9, 9, 9, 9);
    hboxLayout = new QHBoxLayout();
    hboxLayout->setSpacing(6);
    hboxLayout->setContentsMargins(0, 0, 0, 0);
    vboxLayout = new QVBoxLayout();
    vboxLayout->setSpacing(6);
    vboxLayout->setContentsMargins(0, 0, 0, 0);
    spacerItem = new QSpacerItem(20, 2, QSizePolicy::Minimum,
					QSizePolicy::Expanding);

    vboxLayout->addItem(spacerItem);

    QLabel *imageLabel = new QLabel(this);
    imageLabel->setPixmap(pixmap);

    vboxLayout->addWidget(imageLabel);

    spacerItem1 = new QSpacerItem(20, 20, QSizePolicy::Minimum,
					  QSizePolicy::Expanding);

    vboxLayout->addItem(spacerItem1);
    hboxLayout->addLayout(vboxLayout);
    vboxLayout1 = new QVBoxLayout();
    vboxLayout1->setSpacing(6);
    vboxLayout1->setContentsMargins(0, 0, 0, 0);

    int height;
    int width = DEFAULT_EDIT_WIDTH; 

    QLineEdit* lineEdit = NULL;
    QTextEdit* textEdit = NULL;
    if (inputflag && messagecount <= 1) {
	lineEdit = new QLineEdit(this);
	if (messagecount == 1)
	    lineEdit->setText(tr(messages[0]));
	height = lineEdit->font().pointSize() + 4;
	if (height < 0)
	    height = lineEdit->font().pixelSize() + 4;
	if (height < 0)
	    height = lineEdit->heightForWidth(width) + 4;
	lineEdit->setSizePolicy(QSizePolicy::MinimumExpanding,
				QSizePolicy::Fixed);
	lineEdit->setMinimumSize(QSize(width, height));
	lineEdit->setGeometry(QRect(0, 0, width, height));
	vboxLayout1->addWidget(lineEdit);
    }
    else {
	QFont	fixed("monospace");
	fixed.setStyleHint(QFont::TypeWriter);

	textEdit = new QTextEdit(this);
	textEdit->setFont(fixed);
	textEdit->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
	if (nosliderflag)
	    textEdit->setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
	else if (usesliderflag)
	    textEdit->setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOn);
	else
	    textEdit->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
	for (int m = 0; m < messagecount; m++)
	    textEdit->append(tr(messages[m]));
	if (inputflag)
	    textEdit->setReadOnly(false);
	else {
	    textEdit->setLineWidth(1);
	    textEdit->setFrameStyle(noframeflag ? QFrame::NoFrame :
				    QFrame::Box | QFrame::Sunken);
	    textEdit->setReadOnly(true);
	}
	if (usesliderflag)
	    height = DEFAULT_EDIT_HEIGHT;
	else {
	    height = textEdit->font().pointSize() + 4;
	    if (height < 0)
		height = textEdit->font().pixelSize() + 4;
	    if (height < 0)
	        height = textEdit->heightForWidth(width) + 4;
	    height *= messagecount;
	}
	textEdit->setMinimumSize(QSize(width, height));
	textEdit->setSizePolicy(QSizePolicy::MinimumExpanding,
				QSizePolicy::MinimumExpanding);
	textEdit->moveCursor(QTextCursor::Start);
	textEdit->ensureCursorVisible();
	vboxLayout1->addWidget(textEdit);
    }

    hboxLayout1 = new QHBoxLayout();
    hboxLayout1->setSpacing(6);
    hboxLayout1->setContentsMargins(0, 0, 0, 0);
    spacerItem2 = new QSpacerItem(40, 20, QSizePolicy::Expanding,
					  QSizePolicy::Minimum);
    hboxLayout1->addItem(spacerItem2);

    for (int i = 0; i < buttoncount; i++) {
	QueryButton *button = new QueryButton(printflag, this);
	button->setMinimumSize(QSize(72, 32));
	button->setDefault(buttons[i] == defaultbutton);
	button->setQuery(this);
	button->setText(tr(buttons[i]));
	button->setStatus(statusi[i]);
	if (inputflag && buttons[i] == defaultbutton) {
	    if (textEdit) 
		button->setEditor(textEdit);
	    else if (lineEdit) {
		button->setEditor(lineEdit);
		if (buttons[i] == defaultbutton)
		    connect(lineEdit, SIGNAL(returnPressed()),
			    button, SLOT(click()));
	    }
	}
	connect(button, SIGNAL(clicked()), this, SLOT(buttonClicked()));
	hboxLayout1->addWidget(button);
    }

    vboxLayout1->addLayout(hboxLayout1);
    hboxLayout->addLayout(vboxLayout1);
    gridLayout->addLayout(hboxLayout, 0, 0, 1, 1);
    gridLayout->setSizeConstraint(QLayout::SetFixedSize);

    if (inputflag && messagecount <= 1)
	resize(QSize(320, 83));
    else
	resize(QSize(320, 132));

    if (timeout)
	startTimer(timeout * 1000);

    if (exclusiveflag)
	setWindowModality(Qt::WindowModal);
}
